From f74baaaad9d2116ffa9bb35d07502ddea2399bec Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Sun, 25 Apr 2021 16:32:41 -0500
Subject: [PATCH 084/196] Add Log2FixedPoint (#152)

This allows computing the log2 using deterministic integer arithmetic.
Required for making block reward equal to log(difficulty).
---
 src/CMakeLists.txt         |  1 +
 src/test/CMakeLists.txt    |  1 +
 src/test/intmath_tests.cpp | 62 ++++++++++++++++++++++++++++++++++++++
 src/util/intmath.cpp       | 49 ++++++++++++++++++++++++++++++
 src/util/intmath.h         | 14 +++++++++
 5 files changed, 127 insertions(+)
 create mode 100644 src/test/intmath_tests.cpp
 create mode 100644 src/util/intmath.cpp
 create mode 100644 src/util/intmath.h

diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index db355dbae..99619435c 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -375,6 +375,7 @@ add_library(util
 	util/bip32.cpp
 	util/bytevectorhash.cpp
 	util/error.cpp
+	util/intmath.cpp
 	util/message.cpp
 	util/moneystr.cpp
 	util/settings.cpp
diff --git a/src/test/CMakeLists.txt b/src/test/CMakeLists.txt
index 01b0df45d..06acb8018 100644
--- a/src/test/CMakeLists.txt
+++ b/src/test/CMakeLists.txt
@@ -160,6 +160,7 @@ add_boost_unit_tests_to_suite(bitcoin test_bitcoin
 		getarg_tests.cpp
 		hash_tests.cpp
 		interfaces_tests.cpp
+		intmath_tests.cpp
 		inv_tests.cpp
 		key_io_tests.cpp
 		key_tests.cpp
diff --git a/src/test/intmath_tests.cpp b/src/test/intmath_tests.cpp
new file mode 100644
index 000000000..905f99941
--- /dev/null
+++ b/src/test/intmath_tests.cpp
@@ -0,0 +1,62 @@
+#include <util/intmath.h>
+
+#include <cmath>
+
+#include <boost/test/unit_test.hpp>
+#include <test/util/setup_common.h>
+
+BOOST_FIXTURE_TEST_SUITE(intmath_tests, BasicTestingSetup)
+
+BOOST_AUTO_TEST_CASE(log2fixedpoint_tests) {
+    struct TestCase {
+        uint32_t x;
+        size_t prec;
+        int32_t expect;
+        double error;
+    };
+    std::vector<TestCase> test_cases = {
+        {.x = 256, .prec = 1, .expect = 14, .error = 0.0},
+        {.x = 512, .prec = 2, .expect = 28, .error = 0.0},
+        {.x = 1024, .prec = 3, .expect = 56, .error = 0.0},
+        {.x = 1, .prec = 1, .expect = -2, .error = 0.0},
+        {.x = 2, .prec = 1, .expect = 0, .error = 0.0},
+        {.x = 3, .prec = 1, .expect = 1, .error = 0.09},
+        {.x = 4, .prec = 1, .expect = 2, .error = 0.0},
+        {.x = 1, .prec = 2, .expect = -8, .error = 0.0},
+        {.x = 2, .prec = 2, .expect = -4, .error = 0.0},
+        {.x = 3, .prec = 2, .expect = -2, .error = 0.09},
+        {.x = 4, .prec = 2, .expect = 0, .error = 0.0},
+        {.x = 1, .prec = 3, .expect = -24, .error = 0.0},
+        {.x = 2, .prec = 3, .expect = -16, .error = 0.0},
+        {.x = 3, .prec = 3, .expect = -12, .error = 0.09},
+        {.x = 4, .prec = 3, .expect = -8, .error = 0.0},
+        {.x = 1'000, .prec = 8, .expect = 502, .error = 0.01},
+        {.x = 10'000, .prec = 8, .expect = 1352, .error = 0.01},
+        {.x = 100'000, .prec = 8, .expect = 2203, .error = 0.01},
+        {.x = 1'000'000, .prec = 8, .expect = 3053, .error = 0.01},
+        {.x = 10'000'000, .prec = 8, .expect = 3904, .error = 0.01},
+        {.x = 100'000'000, .prec = 8, .expect = 4754, .error = 0.01},
+        {.x = 1'000'000'000, .prec = 8, .expect = 5604, .error = 0.01},
+        {.x = 0x7fff'ffff, .prec = 8, .expect = 5887, .error = 0.01},
+        {.x = 0xffff'ffffU, .prec = 8, .expect = 6143, .error = 0.01},
+        {.x = 1'000, .prec = 16, .expect = -395459, .error = 0.0001},
+        {.x = 10'000, .prec = 16, .expect = -177753, .error = 0.0001},
+        {.x = 100'000, .prec = 16, .expect = 39952, .error = 0.0001},
+        {.x = 1'000'000, .prec = 16, .expect = 257658, .error = 0.0001},
+        {.x = 10'000'000, .prec = 16, .expect = 475364, .error = 0.0001},
+        {.x = 100'000'000, .prec = 16, .expect = 693070, .error = 0.0001},
+        {.x = 1'000'000'000, .prec = 16, .expect = 910776, .error = 0.0001},
+        {.x = 0x7fff'ffff, .prec = 16, .expect = 983039, .error = 0.0001},
+        {.x = 0xffff'ffffU, .prec = 16, .expect = 1048575, .error = 0.0001},
+    };
+    for (auto &test_case : test_cases) {
+        int32_t log = Log2FixedPoint(test_case.x, test_case.prec);
+        double factor = 1U << test_case.prec;
+        double actual = log / factor;
+        double expected = log2(test_case.x / factor);
+        BOOST_CHECK(abs(actual - expected) <= test_case.error);
+        BOOST_CHECK_EQUAL(log, test_case.expect);
+    }
+}
+
+BOOST_AUTO_TEST_SUITE_END()
diff --git a/src/util/intmath.cpp b/src/util/intmath.cpp
new file mode 100644
index 000000000..0f7909fa0
--- /dev/null
+++ b/src/util/intmath.cpp
@@ -0,0 +1,49 @@
+// Copyright (c) 2021 Logos Foundation
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <util/intmath.h>
+
+#include <assert.h>
+
+/**
+ * Computes the log2 of `x`, where `x` is a fixed point number with `precision`
+ * number of binary digits after the fractional point. The result is also such a
+ * fixed point number with the same precision.
+ *
+ * See C. S. Turner, "A Fast Binary Logarithm Algorithm", IEEE Signal Processing
+ * Mag., pp. 124,140, Sep. 2010.
+ */
+int32_t Log2FixedPoint(uint32_t x, const size_t precision) {
+    int32_t b = 1U << (precision - 1);
+    int32_t y = 0;
+
+    assert(precision >= 1 && precision <= 31);
+
+    if (x == 0) {
+        return std::numeric_limits<int32_t>::min(); // negative infinity
+    }
+
+    while (x < 1U << precision) {
+        x <<= 1;
+        y -= 1U << precision;
+    }
+
+    while (x >= 2U << precision) {
+        x >>= 1;
+        y += 1U << precision;
+    }
+
+    uint64_t z = x;
+
+    for (size_t i = 0; i < precision; i++) {
+        z = (z * z) >> precision;
+        if (z >= 2U << precision) {
+            z >>= 1;
+            y += b;
+        }
+        b >>= 1;
+    }
+
+    return y;
+}
diff --git a/src/util/intmath.h b/src/util/intmath.h
new file mode 100644
index 000000000..a7be8c502
--- /dev/null
+++ b/src/util/intmath.h
@@ -0,0 +1,14 @@
+// Copyright (c) 2021 Logos Foundation
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_UTIL_INTMATH_H
+#define BITCOIN_UTIL_INTMATH_H
+
+#include <cstddef>
+#include <cstdint>
+#include <limits>
+
+int32_t Log2FixedPoint(uint32_t x, const size_t precision);
+
+#endif // BITCOIN_UTIL_INTMATH_H
-- 
2.39.5 (Apple Git-154)

